### PR TITLE
[Geo] Integrate Lucene's LatLonShape (BKD Backed GeoShapes) as default `geo_shape` indexing approach

### PR DESCRIPTION
**Pull request by nknize**
This PR exposes lucene's `LatLonShape` field as the default type in `GeoShapeFieldMapper`. To use the new indexing approach, simply set `"type" : "geo_shape"` in the mappings without setting any of the `strategy`, `precision`, `tree_levels`, or `distance_error_pct` parameters. Note the following when using the new indexing approach:

* `geo_shape` query does not support querying by `MULTIPOINT`.
* `LINESTRING` and `MULTILINESTRING` queries do not yet support `WITHIN` relation. This will be added in the future as a `buffered_query`.
* `CONTAINS` relation is not yet supported.
* The `tree`, `precision`, `tree_levels`, `distance_error_pct`, and `points_only` parameters are deprecated.
* This PR will be backported to 6.6 (with deprecation logging included) and a separate parameter removal PR will be opened for 7.0.

All other `geo_shape` features are supported.

### TODO
- [x] Implement `POINT` queries as a `BoundingBox` query with `minLat == maxLat` && `minLon == maxLon`
- [x] Implement `GEOMETRYCOLLECTION` query as a `BooleanQuery`
- [x] Fix any remaining failing tests
- [x] Add deprecation logging for `tree`, `precision`, `tree_levels`, `distance_error_pct`, and `points_only` parameters in this PR (to be backported to 6.x)
- [x] Update docs

relates #32039

### CODE REVIEW

**Comment by nknize** [comment]
This is a wip so I expect some yaml tests (or two) to fail but it is close. I wanted to get it opened to get some eyes, discussion, and iterations on it.

File: server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java
Comment (iverase) [inline server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: I think we can offer a point query by reusing the Box query. We just need to make minLat == maxLat and minLon == maxLon.

Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: :+1:  I'll run some tests with that approach.  I was also considering implementing `GeometryCollection` query as a `BooleanQuery`; though we'll likely need to run some benchmarks to make sure performance is acceptable.

Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: Implemented this for `POINT` queries, and it will work quite nicely. Unfortunately I found (and fixed) a small bug in `LatLonShapeBoundingBoxQuery`. So I'll either need to upgrade the lucene version to commit the implementation or hold off until another PR.
I do not want to implement this approach for `MULTIPOINT` queries, however. It would require rewriting to a `BooleanQuery` which would be quite slow, and abusive, for even just a few points.

**Comment by nknize** [comment]
retest this please

**Comment by nknize** [comment]
Still trying to figure out why these doc changes are failing. @nik9000 any thoughts?

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: why are you making a copy instead of just `List<IndexableField> flist = Arrays.asList(fields)`?

File: server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: should we fail if the length of the `double[]` is not 2?

File: server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: let's wrap in a ConstantScoreQuery?

**Review by jpountz — COMMENTED** [review]
I like the approach. I'm wondering that things would be a bit easier to read eg. if we renamed the current field mapper to something like LegacyGeoShapeFieldMapper, created a new (clean) GeoShapeFieldMapper that uses points, and made the TypeParser of `GeoShapeFieldMapper` delegate to `LegacyGeoShapeFieldMapper.TypeParser` if `strategy` is set and different from `vector`? My motivation is that then the new field mapper would be easier to read and not cluttered with backward compatibility logic?

File: server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: Its wrapped in a `ConstantScoreQuery` on line 432.

File: server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: :+1:

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: Because `createFieldNamesField` adds elements to the list, and Lists returned by `Arrays.asList` cannot be structurally modified.

File: server/src/main/java/org/elasticsearch/common/geo/SpatialStrategy.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/common/geo/SpatialStrategy.java]: ¯\\_(ツ)_/¯     :-)  Will fix.

File: server/src/main/java/org/elasticsearch/common/geo/ShapeRelation.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/common/geo/ShapeRelation.java]: :+1:

**Comment by nknize** [comment]
> I'm wondering that things would be a bit easier to read eg. if we renamed the current field mapper to something like LegacyGeoShapeFieldMapper

:+1: I did it this way when I migrated `geo_point` to `LatLonPoint`. It works nicely and given the discussion from the other day re: versioning and release timing I think its the right way to go.

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (nknize) [reply docs/reference/mapping/types/geo-shape.asciidoc]: :+1:

**Comment by nknize** [comment]
OK... this took a little bit of work but I have made the following changes:

1. refactored `GeoShapeFieldMapper` into:
   -- `LegacyGeoShapeFieldMapper` - provides all legacy Prefix Tree parameters, and uses legacy PrefixTree indexing provided by lucene's `spatial-extras` module
   -- `GeoShapeFieldMapper` - provides new vector based indexing approach provided by lucene's new `LatLonShape` field type
   -- `BaseGeoShapeFieldMapper` - provides support for common parameters used between `GeoShapeFieldMapper` and `LegacyGeoShapeFieldMapper` (e.g., `coerce`, `ignore_malformed`, `ignore_z_value`, `orientation`)

2. Removed the new `vector` strategy
   -- Since `strategy` is only needed for prefix trees and the default indexing approach for newly created indexes is `LatLonShape` there is no need to have users explicitly specify a `strategy` the ideal mapping from here forward is:

```
"properties" : {
  "location" : {
    "type" : "geo_shape"
  }
}
```
   No other parameters are necessary

3. updated documentation to remove vector strategy and better describe the new indexing approach

**Comment by nknize** [comment]
This commit is over +3500/-1500 changes. I'll look at splitting into two separate PRs:

1. Adding support to `ShapeBuilders` for building the Lucene Geometry needed by the new `LatLonShape` indexing approach
2. Refactoring `GeoShapeFieldMapper` and updating `GeoShapeQuery`

**Comment by jpountz** [comment]
That would be great!

**Comment by nknize** [comment]
ShapeBuilder PR is open at #35707. Once it is merged I'll open up the second PR for refactoring `GeoShapeFieldMapper`

**Comment by nknize** [comment]
Merged #35707 into this branch so I don't think there's a need to open a separate PR. This one has been updated with all the ShapeBuilder changes and includes just the Mapper, Query Builder, and all test changes.

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (jpountz) [inline docs/reference/mapping/types/geo-shape.asciidoc]: we need a more qualified name since this links may be used from anywhere in the docs, eg. `geo-shape-indexing-approach`

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (jpountz) [inline docs/reference/mapping/types/geo-shape.asciidoc]: maybe we could mention that indexing performance mostly depends on the number of edges of the input (multi)polygon

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (jpountz) [inline docs/reference/mapping/types/geo-shape.asciidoc]: we need a new line for `vector`?

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (jpountz) [inline docs/reference/mapping/types/geo-shape.asciidoc]: Isn't it `1e-7`? (`360 / 2^32` ~= `8.4e-8` ~= `1e-7`)

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (jpountz) [inline docs/reference/mapping/types/geo-shape.asciidoc]: is it 1e-7 here too?

File: docs/reference/migration/migrate_7_0/mappings.asciidoc
Comment (jpountz) [inline docs/reference/migration/migrate_7_0/mappings.asciidoc]: these changes should be in `docs/reference/migration/migrate_6_6.asciidoc` instead? (only exists on the 6.x branch)

File: server/src/main/java/org/elasticsearch/common/geo/builders/GeometryCollectionBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/geo/builders/GeometryCollectionBuilder.java]: why did you remove this?

File: server/src/main/java/org/elasticsearch/index/mapper/BaseGeoShapeFieldMapper.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/BaseGeoShapeFieldMapper.java]: Why do we hold the orientation here while it was only in the field type previously?

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: getClass().getName() will not be meaningful eg. for an anonymous class, let's just rely on the toString representation of the class?

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java
Comment (jpountz) [reply server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: having a root class doesn't make much sense from a Lucene perspective since you're not expected to manipulate them in a generic way. Maybe we could have a wrapper on the Elasticsearch side to add type safety that would look something like this? Would it address your concern?

```java
public class GeoShape {

  // exactly one of them is non-null
  public final double[] point;
  public final Line line;
  public final Polygon polygon;
  // other types we want to support

  public GeoShape(double[] point) {
    this.point = point;
    this.line = null;
    this.polygon = null;
  }

  public GeoShape(Line line) {
    this.point = null;
    this.line = line;
    this.polygon = null;
  }

  public GeoShape(Polygon polygon) {
    this.point = null;
    this.line = null;
    this.polygon = polygon;
  }
}
```

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java
Comment (jpountz) [reply server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: thanks I had missed this

File: server/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: should we test with the non-legacy mapper instead?

File: server/src/test/java/org/elasticsearch/index/mapper/ExternalMapper.java
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/index/mapper/ExternalMapper.java]: let's use the non-legacy mapper here too

File: server/src/main/java/org/elasticsearch/index/mapper/BaseGeoShapeFieldMapper.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/BaseGeoShapeFieldMapper.java]: do we want to disallow them on 7.0+ indices? I'm happy with it being done in a follow-up PR, just asking.

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java
Comment (imotov) [reply server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: I think this will just replace this huge `if (shape instanceof ... )` with `if (shape != null)` switch.

Comment (jpountz) [reply server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: Indeed. I was trying to see whether that would address your concern that there is no root geometry class.

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (nknize) [reply docs/reference/mapping/types/geo-shape.asciidoc]: No. I need to change "three" to "two".  I removed the vector option altogether and just set `LatLonShape` as the default indexing strategy. This way users have to explicitly specify `tree` or `strategy` parameter to use the legacy (soon to be removed) PrefixTree approach. I thought this was cleaner.

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (nknize) [reply docs/reference/mapping/types/geo-shape.asciidoc]: :+1:

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (nknize) [reply docs/reference/mapping/types/geo-shape.asciidoc]: :+1:

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (nknize) [reply docs/reference/mapping/types/geo-shape.asciidoc]: :+1:

File: server/src/test/java/org/elasticsearch/common/geo/GeoJsonShapeParserTests.java
Comment (nknize) [reply server/src/test/java/org/elasticsearch/common/geo/GeoJsonShapeParserTests.java]: :+1:

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (nknize) [reply docs/reference/mapping/types/geo-shape.asciidoc]: No. I went ahead and removed `vector` altogether and made BKD indexing the default. No `strategy` needed. Its now been deprecated along with all other PrefixTree specific parameters. Just made more sense.

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (nknize) [reply docs/reference/mapping/types/geo-shape.asciidoc]: :+1:  That's right... I was wrong.

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (nknize) [reply docs/reference/mapping/types/geo-shape.asciidoc]: :+1:

File: docs/reference/migration/migrate_7_0/mappings.asciidoc
Comment (nknize) [reply docs/reference/migration/migrate_7_0/mappings.asciidoc]: Yes I agree.... I can do that in the backport? Or should I open a whole new PR to handle that?

File: server/src/main/java/org/elasticsearch/common/geo/builders/GeometryCollectionBuilder.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/common/geo/builders/GeometryCollectionBuilder.java]: Because for Lucene geometry the shape could be a point defined as a `double[]` which fails downstream casting.

File: server/src/main/java/org/elasticsearch/index/mapper/BaseGeoShapeFieldMapper.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/mapper/BaseGeoShapeFieldMapper.java]: Because orientation is used to build the geometry regardless of whether its S4J or Lucene. So it made more sense to have the parameter be part of the builder.

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: :+1:

File: server/src/main/java/org/elasticsearch/index/mapper/BaseGeoShapeFieldMapper.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/mapper/BaseGeoShapeFieldMapper.java]: :+1:  Yes. I was going to do this in a separate PR to simplify backporting these changes to 6.6

File: server/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java
Comment (nknize) [reply server/src/test/java/org/elasticsearch/common/geo/GeoWKTShapeParserTests.java]: :+1:

File: server/src/test/java/org/elasticsearch/index/mapper/ExternalMapper.java
Comment (nknize) [reply server/src/test/java/org/elasticsearch/index/mapper/ExternalMapper.java]: I have this there for backcompat testing. The `build()` method checks index version and uses this for indexes created before 6.6

**Comment by nknize** [comment]
I think this is ready for another round of review. Just note that `vector` is no longer a strategy. It made more sense to make BKD the new default indexing option and PrefixTrees an "opt in" strategy. Also, `strategy` is a deprecated option since it's only relevant for Prefix Tree indexing. It didn't make much sense to add a new strategy to a soon to be removed parameter.

File: server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: @imotov perhaps this can be resolved in a future PR by using the coming Geo API?

Comment (imotov) [reply server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java]: :+1:

File: docs/reference/mapping/types/geo-shape.asciidoc
Comment (nknize) [reply docs/reference/mapping/types/geo-shape.asciidoc]: :+1:

**Comment by nknize** [comment]
retest this please

File: server/src/main/java/org/elasticsearch/index/mapper/LegacyGeoShapeFieldMapper.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/LegacyGeoShapeFieldMapper.java]: let's always have the strategy in the xcontent for legacy shapes to avoid any confusion? ie. remove the if statement and always execute the body

File: server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: it's a bit weird to recommend deprecated functionality?

File: server/src/test/java/org/elasticsearch/index/mapper/ExternalMapper.java
Comment (jpountz) [reply server/src/test/java/org/elasticsearch/index/mapper/ExternalMapper.java]: ok thanks for explaining

**Review by jpountz — COMMENTED** [review]
Thanks for the iteration @nknize! I left some minor comments.

File: server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java]: lol.. that's a great point. I'll remove the recommendation

File: server/src/main/java/org/elasticsearch/index/mapper/LegacyGeoShapeFieldMapper.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/mapper/LegacyGeoShapeFieldMapper.java]: :+1:  great idea

File: server/src/main/java/org/elasticsearch/common/geo/parsers/GeoJsonParser.java
Comment (nknize) [reply server/src/main/java/org/elasticsearch/common/geo/parsers/GeoJsonParser.java]: :+1:

**Comment by nknize** [comment]
Will keep an eye on the tests... but it looks like this is good to go.

**Review by jpountz — APPROVED** [review]
LGTM

**Comment by philvarner** [comment]
One side effect of the change `Implement GEOMETRYCOLLECTION query as a BooleanQuery` seems to be that large geometry collections (my case, the System for Terrestrial Ecosystem Parameterization (STEP) geojson) that, prior to 6.6, were counted as a single clause, now seem to have each geometry count as a clause, and will hit the maxClauseCount default limit of 1024.  This was unexpected and something I couldn't find documented anywhere, so I thought it would be useful to at least comment here about it.

**Pull request merged**
